flow/operations.py

```diff
@@ -268,4 +334,144 @@ def operation(job):
+    :type cmd:
+        str
+    :param jobs:
+        The sequence of jobs associated with this operation.
+    :type job:
+        :py:class:`signac.Job`.
+    :param directives:
+        A dictionary of additional parameters that provide instructions on how
+        to execute this operation, e.g., specifically required resources.
+    :type directives:
+        :class:`dict`
+    """
+    MAX_LEN_ID = 100
+
+    def __init__(self, name, cmd, jobs, directives=None):
+        if not len(jobs):
+            raise ValueError("The jobs argument cannot be empty!")
+        self.name = name
+        self.jobs = jobs
+        self.cmd = cmd
+        if directives is None:
+            directives = dict()  # default argument
+        else:
+            directives = dict(directives)  # explicit copy
+
+        # Keys which were explicitly set by the user, but are not evaluated by the
+        # template engine are cause for concern and might hint at a bug in the template
+        # script or ill-defined directives. We are therefore keeping track of all
+        # keys set by the user and check whether they have been evaluated by the template
+        # script engine later.
+        keys_set_by_user = set(directives.keys())
+
+        # Evaluate strings and callables for job:
+        def evaluate(value):
+            if value and callable(value):
+                return value(* self.jobs)
+            elif isinstance(value, six.string_types):
+                if len(jobs) == 1:
+                    return value.format(job=self.jobs[0])
+                else:
+                    return value.format(jobs=self.jobs)
+            else:
+                return value
+
+        directives.setdefault('np',
+                              evaluate(directives.get('nranks', 1))
+                              * evaluate(directives.get('omp_num_threads', 1)))
+        directives.setdefault('ngpu', 0)
+        directives.setdefault('nranks', 0)
+        directives.setdefault('omp_num_threads', 0)
+
+        # We use a special dictionary that allows us to track all keys that have been
+        # evaluated by the template engine and compare them to those explicitly set
+        # by the user. See also comment above.
+        self.directives = TrackGetItemDict(
+            {key: evaluate(value) for key, value in directives.items()})
+        self.directives._keys_set_by_user = keys_set_by_user
+
+    def __str__(self):
+        return "{}({})".format(self.name, ', '.join([job.get_id() for job in self.jobs]))
+
+    def __repr__(self):
+        return "{type}(name='{name}', cmd={cmd}, jobs='{jobs}', directives={directives})".format(
+            type=type(self).__name__,
+            name=self.name,
+            jobs=', '.join([job.get_id() for job in self.jobs]),
+            cmd=repr(self.cmd),
+            directives=self.directives)
+
+    @property
+    def job(self):
+        assert len(self.jobs) >= 1
+        return self.jobs[0]
```

Comment:
Can we guarantee that the `JobsOperation` is always constructed with jobs in the same order? Since status is stored on a project level this probably doesn't affect any internals, but I recall that we maybe discussed what were sensible choices here.